presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/StreamPropertyDerivations.java

```diff
@@ -0,0 +1,566 @@
+
+            Optional<Set<Symbol>> partitionSymbols = layout.getPartitioningColumns()
+                    .flatMap(columns -> getNonConstantSymbols(columns, assignments, constants));
+
+            return new StreamProperties(MULTIPLE, partitionSymbols, false);
+        }
+
+        private static Optional<Set<Symbol>> getNonConstantSymbols(Set<ColumnHandle> columnHandles, Map<ColumnHandle, Symbol> assignments, Set<ColumnHandle> globalConstants)
+        {
+            // Strip off the constants from the partitioning columns (since those are not required for translation)
+            Set<ColumnHandle> constantsStrippedPartitionColumns = columnHandles.stream()
+                    .filter(column -> !globalConstants.contains(column))
+                    .collect(toImmutableSet());
+            ImmutableSet.Builder<Symbol> builder = ImmutableSet.builder();
+
+            for (ColumnHandle column : constantsStrippedPartitionColumns) {
+                Symbol translated = assignments.get(column);
+                if (translated == null) {
+                    return Optional.empty();
+                }
+                builder.add(translated);
+            }
+
+            return Optional.of(builder.build());
+        }
+
+        @Override
+        public StreamProperties visitExchange(ExchangeNode node, List<StreamProperties> inputProperties)
+        {
+            // remote exchange always produces a single stream
+            if (node.getScope() == REMOTE) {
+                return StreamProperties.singleStream();
+            }
+
+            switch (node.getType()) {
+                case GATHER:
+                    return StreamProperties.singleStream();
+                case REPARTITION:
+                    if (node.getPartitionFunction().getPartitioningHandle().equals(FIXED_RANDOM_DISTRIBUTION)) {
+                        return new StreamProperties(FIXED, Optional.empty(), false);
+                    }
+                    return new StreamProperties(
+                            FIXED,
+                            Optional.of(
+                            node.getPartitionFunction().getPartitionFunctionArguments().stream()
+                                    .map(PartitionFunctionArgumentBinding::getColumn)
+                                    .collect(toImmutableList())),
+                            false);
+                case REPLICATE:
+                    return new StreamProperties(MULTIPLE, Optional.empty(), false);
+            }
+
+            throw new UnsupportedOperationException("not yet implemented");
+        }
+
+        //
+        // Nodes that rewrite and/or drop symbols
+        //
+
+        @Override
+        public StreamProperties visitProject(ProjectNode node, List<StreamProperties> inputProperties)
+        {
+            StreamProperties properties = Iterables.getOnlyElement(inputProperties);
+
+            // Only identity projected symbols are passed through
```

Comment:
I might rephrase the comment as "We can describe properties in terms of inputs that are projected unmodified"